### PR TITLE
Improve performance of array add  operation and type checker

### PR DESCRIPTION
**Pull request by warunalakshitha**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/43694

- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by heshanpadmasiri — APPROVED** [review]
LGTM

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 1]
```diff
@@ -257,12 +257,15 @@ public static boolean anyToJBoolean(Object sourceVal) {
      * @return true if the value belongs to the given type, false otherwise
      */
     public static boolean checkIsType(Object sourceVal, Type targetType) {
-        Context cx = context();
         Type sourceType = getType(sourceVal);
+        if (sourceType == targetType || (sourceType.getTag() == targetType.getTag() && sourceType.equals(targetType))) {
```
Comment (heshanpadmasiri) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:261]: I think with this change we no longer this equals check in TypeCheckCache https://github.com/ballerina-platform/ballerina-lang/blob/c4fe578f0af44a71df17d439d4fe0999b6bee248/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/types/semtype/TypeCheckCache.java#L32

**Review by HindujaB — APPROVED** [review]
LGTM

**Review by lochana-chathura — APPROVED** [review]
LGTM

**Pull request merged**